public/API/API_GetFollowedUsersList.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+/*
+ *  API_GetFollowedUsersList - returns list of the caller's followed users
+ *    o : offset - number of entries to skip (default: 0)
+ *    c : count - number of entries to return (default: 100, max: 500)
+ *  int         Count                       number of user records returned in the response
+ *  int         Total                       number of user records the user actually has overall
+ *  array       Results
+ *   object      [value]
+ *    string     User                       username
+ *    int        Points                     number of hardcore points the user has earned
+ *    int        PointsSoftcore             number of softcore points the user has earned
+ *    boolean    FollowsBack                whether the followed user follows the caller user back
+ */
+
+use App\Models\User;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Validator;
+
+$input = Validator::validate(Arr::wrap(request()->query()), [
+    'c' => 'nullable|integer|min:0',
+    'o' => 'nullable|integer|min:1|max:500',
+]);
+
+$offset = $input['o'] ?? 0;
+$count = $input['c'] ?? 100;
+
+/** @var User $user */
+$user = Auth::user();
+
+$totalUsers = $user->followedUsers()
+    ->whereNull('Deleted')
+    ->count();
+
+$usersList = $user->followedUsers()
+    ->whereNull('Deleted')
+    ->orderBy('LastActivityID', 'DESC')
+    ->skip($offset)
+    ->take($count)
+    ->get()
+    ->map(function ($followedUser) use ($user) {
+        return [
+            'User' => $followedUser->display_name,
+            'Points' => $followedUser->points,
+            'PointsSoftcore' => $followedUser->points_softcore,
+            'FollowsBack' => $followedUser->isFollowing($user),
```

Comment:
There's an N+1 query problem here. If the endpoint returns 100 results, it's going to initiate 100+ database calls.

Reply:
Came up with a solution, might not be the most elegant but it does do the job of avoiding the N+1 queries to fetch this data. With the new implementation it's just two queries, one for the total and another one for the filtered list.

Looking forward to your thoughts :)

Reply:
Hey, thanks for the suggestion, did tried several ways but unfortunately none of them kept the logic working without actually doing a raw SQL. Feel free to guide me a bit more if you have any idea.

I did removed the shareable code from `ActsAsCommunityMember` tho.

Reply:
No worries. How about something like this?:
```php
$usersList = $user
  ->followedUsers()
  ->whereNull("Deleted")
  ->with([
    "relatedUsers" => fn($q) => $q
      ->select("UserAccounts.ID", "related_user_id")
      ->where("related_user_id", $user->id)
      ->withPivot("Friendship")
  ])
  ->orderBy("LastActivityID", "DESC")
  ->skip($offset)
  ->take($count)
  ->get()
  ->map(
    fn($followedUser) => [
      "User" => $followedUser->display_name,
      "Points" => $followedUser->points,
      "PointsSoftcore" => $followedUser->points_softcore,
      "FollowsBack" =>
        $followedUser->relatedUsers->first()?->pivot?->Friendship ===
        UserRelationship::Following
    ]
  );
```

This cuts the query count down to 3, which is much more palatable, though I have not checked the data for accuracy :)

Reply:
This last suggestion looks interesting, I can give it a try if you'd like, but perhaps is not a bad idea to use the DB::raw solution which keeps the total query count on 2.

A bit more lines, but fetches the data in a single query.
```
$userTable = $user->getTable();
$friendTable = (new UserRelation())->getTable();
$friendshipStatus = UserRelationship::Following;

$usersList = $user->followerUsers()
    ->select([
        "$userTable.*",
        DB::raw("EXISTS (
            SELECT 1
            FROM $friendTable AS ur
            WHERE ur.user_id = $friendTable.related_user_id
            AND ur.related_user_id = $friendTable.user_id
            AND ur.Friendship = $friendshipStatus
        ) AS am_i_following_back"),
        "$friendTable.user_id as pivot_user_id",
        "$friendTable.related_user_id as pivot_related_user_id",
        "$friendTable.Friendship as pivot_Friendship",
    ])
    ->whereNull('Deleted')
    ->orderBy('LastActivityID', 'DESC')
    ->skip($offset)
    ->take($count)
    ->get()
    ->map(function ($followerUser) {
        return [
            'User' => $followerUser->display_name,
            'Points' => $followerUser->points,
            'PointsSoftcore' => $followerUser->points_softcore,
            'AmIFollowing' => filter_var($followerUser->am_i_following_back, FILTER_VALIDATE_BOOLEAN),
        ];
    });
```

What do you think?

Reply:
I lean towards the query proposed in https://github.com/RetroAchievements/RAWeb/pull/2887#discussion_r1864254157 :-)

The reasons why are:
* Using Eloquent relationships is a bit more idiomatic with respect to Laravel than reaching for a `DB::raw()`.
* Our database is still frequently changing. Using `DB::raw()` is great for optimizing performance in a maximal way (which you've done!), but in this case it may be a premature optimization. When underlying table structures change, ideally these queries will continue working because they're using Eloquent's fluent syntax.

Reply:
That really does make sense, thanks for taking the time to share your point. I have tested your code snippet and worked perfectly, now code-wisely looks much prettier :)